expression/new_constant_propagation.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"bytes"
+
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// exprSet is a Set container for expressions, each expression in it is unique.
+// `tombstone` is deleted mark, if tombstone[i] is false, data[i] is invalid.
+// `index` use expr.HashCode() as key.
+type exprSet struct {
+	data      []Expression
+	tombstone []bool
+	index     map[string]struct{}
+}
+
+func (s *exprSet) Append(e Expression) bool {
+	if _, ok := s.index[string(e.HashCode(nil))]; ok {
+		return false
+	}
+
+	s.data = append(s.data, e)
+	s.tombstone = append(s.tombstone, false)
+	s.index[string(e.HashCode(nil))] = struct{}{}
+	return true
+}
+
+func (s *exprSet) Slice() []Expression {
+	idx := 0
+	for i := 0; i < len(s.data); i++ {
+		if !s.tombstone[i] {
+			s.data[idx] = s.data[i]
+			idx++
+		}
+	}
+	return s.data[:idx]
+}
+func (s *exprSet) SetConstFalse() {
+	for i := 0; i < len(s.data); i++ {
+		s.tombstone[i] = true
+	}
+	s.Append(&Constant{
+		Value:   types.NewDatum(false),
+		RetType: types.NewFieldType(mysql.TypeTiny),
+	})
+}
+
+// PropagateConstantNew propagate constant values of deterministic predicates in a condition.
+func PropagateConstantNew(ctx sessionctx.Context, conditions []Expression) []Expression {
+	var exprs exprSet
+	exprs.data = make([]Expression, 0, len(conditions))
+	exprs.tombstone = make([]bool, 0, len(conditions))
+	exprs.index = make(map[string]struct{}, len(conditions))
+	for _, v := range conditions {
+		exprs.Append(v)
+	}
+
+	fixPoint(ctx, &exprs)
+	return exprs.Slice()
+}
+
+// fixPoint is the core of the constant propagation algorithm.
+// It will iterate the expression set over and over again, pick two expressions,
+// apply one to another.
+// If new conditions can be infered, they will be append into the expression set.
+// Until no more conditions can be infered from the set, the algorithm finish.
+func fixPoint(ctx sessionctx.Context, exprs *exprSet) {
+	for {
+		saveLen := len(exprs.data)
+		iterOnce(ctx, exprs)
+		if saveLen == len(exprs.data) {
```

Comment:
Seems `exprs.data` is not modified during `iterOnce`.

Reply:
That slice is append only.